### DISCUSSION TITLE
Name of the ccl executable

### DISCUSSION THREAD

**Discussion by daute** [Clozure/ccl#500] [General]
Is there a reason, why the executable (& image) has a different name for each OS?

E.g. `ccl/dx86cl64` (Apple), `ccl/fx86cl64` (Freebsd), `ccl/lx86cl64` (Linux), ...

It would be much easier, if the name (of the executable and image) would be the same for all OS, especially if CCL is used in other programs (e.g. Maxima).

Best regards, Wolfgang

**Comment by xrme** [comment]
The reason is partly historical. Also, there is (or has been) a development practice where the CCL sources are shared via NFS or the like, and having separate, per-platform names for the lisp kernel and the heap image keeps the various per-platform binaries from overwriting each other. This comes in handy when cross-compiling, too.

Here's one approach that I use. See [scripts/ccl64](https://github.com/Clozure/ccl/blob/master/scripts/ccl64). You can adapt this script as needed, put it somewhere on your path (e.g., `~/bin/ccl`), and that lets you type "ccl" and get the right thing no matter what platform you are on.

**Comment by informatimago** [comment]
@daute I would note that "fat" executables (multiple architectures) are also possible on some platforms, and with different rules and conventions.  For example,  on MacOS, you will have for example:
```
$ file /Applications/Safari.app/Contents/MacOS/Safari
/Applications/Safari.app/Contents/MacOS/Safari: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64] [arm64e:Mach-O 64-bit executable arm64e]
/Applications/Safari.app/Contents/MacOS/Safari (for architecture x86_64):	Mach-O 64-bit executable x86_64
/Applications/Safari.app/Contents/MacOS/Safari (for architecture arm64e):	Mach-O 64-bit executable arm64e
```
Building this fat binary uses platform specific tools (and requires compiling and generating images for those multiple platforms, so at one time we'd have to have on the disk both dx86cl64 and dxarmcl64e and use lipo to merge them into a Closure CL fat ("universal") executable. But other platforms use other conventions and tools.  Hence going thru the ccl script is nice, and it works also via NFS mounted volumes.
(Note: on MacOS, Applications/foo.app/Contents/MacOS/foo can be a shell script, to further dispatch and find an actual executable too).
